llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+
+						//Note that here we return
+						return;
+					}
+					default:
+					{
+						//Default handling of not doing anything is conservative
+						break;
+					}
+				}
+
+				if (next)
+				{
+					//We know where execution should proceed
+					setIncomingBB(I.getParent());
+					getLastStack().CurBB = const_cast<BasicBlock*>(next);
+					getLastStack().CurInst = getLastStack().CurBB->begin();
+					//Also here we have set the proper state for the execution so we can return
+					return;
+				}
+				else
+				{
+					skip = true;
+				}
+			}
+
+			//We are inside a call, here we assume all failure to execute are non-recoverable (as in no information could be gained)
+			if (skip)
+			{
+				//Pop current stack (will happen recursively over the call-stack)
+				popSingleStack();
+				// TODO(carlo): remove also stronglyKnownBits
+				removeFromMaps(getLastStack().Caller);
+				return;
+			}
+		}
+		//If we are here it means we have to actually perform the execution via Interpreter
+
+		//Dispatch to the Interpreter's visitor for the given Instructon
+		Interpreter::visit(const_cast<Instruction&>(I));
+
+		//Add  knownBits information
+		assignToMaps(&I, computeStronglyKnownBits(I));
+	}
+
+	/// Create a new interpreter object.
+	///
+	static ExecutionEngine* create(std::unique_ptr<Module> M, std::string *ErrStr)
+	{
+		// Tell this Module to materialize everything and release the GVMaterializer.
+		if (Error Err = M->materializeAll()) {
+			std::string Msg;
+			handleAllErrors(std::move(Err), [&](ErrorInfoBase &EIB) {
+					Msg = EIB.message();
+					});
+			if (ErrStr)
+				*ErrStr = Msg;
+			// We got an error, just return 0
+			return nullptr;
+		}
+
+		return new PartialInterpreter(std::move(M));
+	}
+};
+
+static void removeEdgeBetweenBlocks(llvm::BasicBlock* from, llvm::BasicBlock* to)
+{
+	//Given two BasicBlocks, remove all edges between them.
+	//Multiple edges are legal and might happen since SwitchInst might have more that one case going to the same BB
+
+	{
+		//First count how many edges there are betwen from and to
+		uint32_t cardinality = 0;
+		for (BasicBlock* bb : predecessors(to))
+			if (bb == from)
+				cardinality++;
+
+		//Then call removePredecessor to fix up the phis of to
+		for (uint32_t i=0; i<cardinality; i++)
+			to->removePredecessor(from);
+	}
+
+	{
+		//Then fix up from's terminator
+		llvm::Instruction* I = from->getTerminator();
+
+		switch (I->getOpcode())
+		{
+			case Instruction::Br:
+			{
+				BranchInst* BI = cast<BranchInst>(I);
+				if (BI->isConditional() && BI->getSuccessor(0) != BI->getSuccessor(1))
+				{
+					//Turn to unconditional
+					llvm::BasicBlock* other = nullptr;
+					if (BI->getSuccessor(0) != to)
+						other = BI->getSuccessor(0);
+					else
+						other = BI->getSuccessor(1);
+					BI->eraseFromParent();
+					BranchInst::Create(other, from);
+				}
+				else
+				{
+					BI->eraseFromParent();
+					new UnreachableInst(from->getParent()->getParent()->getContext(), from);
```

Comment:
Context is available everywhere in the hierarchy, not need for the getParent calls.